### DISCUSSION TITLE
Trying to insert data from a CSV file into Supabase

### DISCUSSION THREAD

**Discussion by jaridley** [supabase/supabase#11869] [Questions]
Hello, I am attempting to insert data from a CSV file into Supabase using a Python program. Has anyone attempted this successfully?

**Comment by aidenkwong** [comment]
1. You can use [pandas](https://pandas.pydata.org/) to load your CSV file and use [psycopg2](https://pypi.org/project/psycopg2/) to connect to the database and Insert data with SQL query. It works just fine if your file isn't too big.

In my case, I have a 4.6 MB CSV file (about 470k rows) and it will take me so long to do so. I've tried using [psql](https://www.postgresguide.com/utilities/psql/) to run COPY query to the table but it said I don't have enough permission. 

> ERROR:  must be superuser or have privileges of the pg_read_server_files role to COPY from a file

2. Then, I used [pgadmin](https://github.com/pgadmin-org/pgadmin4) to import data from CSV and it worked! But remember to make sure your CSV data types exactly match column types in the table.

**Reply by gaslitbytech** [reply]
> ERROR: must be superuser or have privileges of the pg_read_server_files role to COPY from a file

There is a subtle difference betwen `COPY` and `\COPY`.

From the postgres docs: [sql copy](https://www.postgresql.org/docs/current/sql-copy.html)

> Do not confuse `COPY` with the psql instruction [\copy](https://www.postgresql.org/docs/current/app-psql.html#APP-PSQL-META-COMMANDS-COPY). `\copy` invokes `COPY FROM STDIN` or `COPY TO STDOUT`, and then fetches/stores the data in a file accessible to the psql client. Thus, file accessibility and access rights depend on the client rather than the server when `\copy` is used.

On a hosted service like supabase, AWS RDS etc you never going to get ~~superuser~~ access to the filesystem of the database server. Though you can use `\COPY` with `psql` to use your `psql` clients filesystem.

For example I'd run a sql like the following using `psql`.

```sql
  CREATE TEMP TABLE temp_csv (
    "ID"       TEXT,
    "NAME"     TEXT,
    "COMPANY"  TEXT
  );

  \COPY temp_csv("ID","NAME","COMPANY") FROM ./missing-data.csv DELIMITER ',' CSV;
```

**Comment by jaridley** [comment]
Thanks for replying to my email Aiden. I was able to figure out my issue. I
had to keep the data from the csv file in the same format by using
orient='record' in the to_dict function (Ex: to_dict(orient='records').
After that I was able to insert my data.
